### ISSUE TITLE
StringIndexOutOfBoundsException in TypeStringProvider.resolveIfPossible() when referring to enum in to-be-generated value

### ISSUE THREAD

**Issue by bhamiltoncx** [immutables/immutables#286]
Using `javac` 1.8.0_65 and Immutables.org 2.1.14, I'm having trouble generating a value type which refers to an enum in a to-be-generated value type.

Steps to reproduce:

1) Unzip attached `enum-immutables.zip`
2) Run `./build.sh`

Expected result:

Build completes successfully and value types are generated in `gen` subdirectory.

Actual result:

Annotation processor crashes with `StringIndexOutOfBoundsException` in `TypeStringProvider.resolveIfPossible()`:

```
% ./build.sh
+ '[' '!' -d gen ']'
+ mkdir gen
+ javac -classpath jar/value-2.1.14.jar -sourcepath src:gen -processorpath jar/value-2.1.14.jar -proc:only -s gen src/bar/Bar.java src/foo/Foo.java
src/foo/Foo.java:5: error: cannot find symbol
import bar.ImmutableBar;
          ^
  symbol:   class ImmutableBar
  location: package bar
src/foo/Foo.java:9: error: package ImmutableBar does not exist
  public abstract ImmutableBar.Type getType();
                              ^
error: org.immutables.value.internal.$processor$.$Processor threw java.lang.StringIndexOutOfBoundsException: String index out of range: -1
  	at java.lang.String.substring(String.java:1931)
  	at org.immutables.value.internal.$processor$.meta.$TypeStringProvider.resolveIfPossible($TypeStringProvider.java:145)
  	at org.immutables.value.internal.$processor$.meta.$TypeStringProvider.appendResolved($TypeStringProvider.java:127)
  	at org.immutables.value.internal.$processor$.meta.$TypeStringProvider.caseType($TypeStringProvider.java:236)
  	at org.immutables.value.internal.$processor$.meta.$TypeStringProvider.process($TypeStringProvider.java:101)
  	at org.immutables.value.internal.$processor$.meta.$ValueAttribute.initTypeName($ValueAttribute.java:884)
  	at org.immutables.value.internal.$processor$.meta.$ValueAttribute.initAndValidate($ValueAttribute.java:846)
  	at org.immutables.value.internal.$processor$.meta.$AccessorAttributesCollector.collect($AccessorAttributesCollector.java:96)
  	at org.immutables.value.internal.$processor$.meta.$ValueTypeComposer.compose($ValueTypeComposer.java:108)
  	at org.immutables.value.internal.$processor$.meta.$Round.collectValues($Round.java:112)
  	at org.immutables.value.internal.$processor$.$Processor.process($Processor.java:47)
  	at org.immutables.value.internal.$generator$.$AbstractGenerator.process($AbstractGenerator.java:89)
  	at com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:794)
  	at com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:705)
  	at com.sun.tools.javac.processing.JavacProcessingEnvironment.access$1800(JavacProcessingEnvironment.java:91)
  	at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1035)
  	at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1176)
  	at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1170)
  	at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:856)
  	at com.sun.tools.javac.main.Main.compile(Main.java:523)
  	at com.sun.tools.javac.main.Main.compile(Main.java:381)
  	at com.sun.tools.javac.main.Main.compile(Main.java:370)
  	at com.sun.tools.javac.main.Main.compile(Main.java:361)
  	at com.sun.tools.javac.Main.compile(Main.java:56)
  	at com.sun.tools.javac.Main.main(Main.java:42)
3 errors
```

I tried reproducing this with a similar test in `value-fixture` but was unable to (the test passed successfully):

https://github.com/bhamiltoncx/immutables/commit/dae922fb8ece88cd6c778864a6055148cb455d9c

so it's possible I'm invoking the annotation processor incorrectly.

However, this is how Buck invokes the Immutables.org annotation processor, so I'm pretty sure it's OK. 

For context, I'm trying to move all of Buck to the ecj compiler instead of javac, but I'm hitting an error which boils down to a similar issue (any methods which refer to enums from to-be-generated types silently fail to generate) when I do so.

[enum-immutables.zip](https://github.com/immutables/immutables/files/178407/enum-immutables.zip)

**Comment by elucash** [comment]
I'll try to dig into it. Thank you for reporting this!

**Comment by bhamiltoncx** [comment]
Thanks for the speedy reply! FYI, this also reproduces with 2.0.21, so it's not a new bug.

What's confusing to me is Buck already uses the "value type which refers to an enum in a not-yet-generated type" pattern successfully here:

https://github.com/facebook/buck/blob/9e292a4d3dad1743fad46b00fe854b56f769ccb4/src/com/facebook/buck/android/AbstractNdkPlatformTargetConfiguration.java#L40

I think there's some very subtle ordering issue where depending on how many rounds the annotation processor needs to do, referring to an enum via a not-yet-generated class may work or it may not.

**Comment by bhamiltoncx** [comment]
Here's the work I did to start using the Eclipse compiler to invoke the Immutables annotation processor in the Buck build instead of using javac in a separate step:

https://github.com/facebook/buck/commits/immutables-issue-286

The symptom here is that methods in generated classes which refer to enums in to-be-generated types are just missing:

https://gist.github.com/bhamiltoncx/de92dc5716b16cbe9daf

I also tried upgrading to 2.1.14 when I ran into this issue, but it didn't help.

**Comment by elucash** [comment]
Our general strategy with not yet-to-be-generated or otherwise absent types is try to resolve them using source imports or if unable to do it just copy star-imports from source if possible and just copy literally the type as it is in an abstract value. Other strategies like some other annotation processors are using (postponing to later rounds or forcing all types to be resolvable) are unpractical for the _Immutables_. But here we have something that is not working per our strategy and I hope it's fixable. Thank you for the detailed clues, this will help for sure.

**Comment by bhamiltoncx** [comment]
Got it. So in Buck's case, no import should be necessary since the two types are in the same package. The test case is probably related but slightly different.

**Comment by elucash** [comment]
The difference we might have looking for may be the use of enum from the same package vs use from other package, where it worked for the same package, but import processing was broken for types nested in yet-to-be-generated types imported from other packages.
I released the fix release 2.1.15 to evaluate this and some other fixes. I hope this will make to proceed to a next stumbling point (if any) in your endeavor with ECJ. Thank you for reporting this with helpful repros!

**Comment by bhamiltoncx** [comment]
Awesome. This issue is fixed, but the ECJ issue remains. I filed #288 to track that.
